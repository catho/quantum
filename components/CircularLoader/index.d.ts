import { FC } from 'react';

export interface CircularLoaderProps {
    size?: 'medium' | 'large' | 'xlarge' | 'xxlarge' | 'xxxlarge';
    skin?: 'primary' | 'secondary';
}

declare const CircularLoader: FC<CircularLoaderProps>;

export default CircularLoader
