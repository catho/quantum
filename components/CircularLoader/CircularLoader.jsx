import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import { colors, gutter as defaultGutter } from '../shared/theme';

const getSizePixels = (size, gutter) => {
  const sizes = {
    medium: gutter * 2, // 16
    large: gutter * 3, // 24
    xlarge: gutter * 4, // 32
    xxlarge: gutter * 5, // 40
    xxxlarge: gutter * 6, // 48
  };

  return sizes[size];
};

const Content = styled.svg`
  left: 50%;
  overflow: hidden;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);

  ${({ color }) => `
    color: ${color};
  `}
`;

const circularRotate = keyframes`
  100% {
    transform: rotate(360deg);
  }
`;

const circularDash = keyframes`
  0% {
    stroke-dasharray: 1px, 200px;
    stroke-dashoffset: 0px;
  }
  50% {
    stroke-dasharray: 100px, 200px;
    stroke-dashoffset: -15px;
  }
  100% {
    stroke-dasharray: 100px, 200px;
    stroke-dashoffset: -125px;
  }
`;

const Circle = styled.circle`
  animation: ${circularDash} 1.4s ease-in-out infinite;
  cx: 44;
  cy: 44;
  fill: none;
  line-height: 1;
  r: 20.2;
  stroke-dasharray: 80px, 200px;
  stroke-dashoffset: 0px;
  stroke-width: 3.6;
  stroke: currentColor;

  ${({ color }) => `
    color: ${color};
  `}
`;

const Wrapper = styled.div`
  animation: ${circularRotate} 1.4s linear infinite;
  position: relative;

  ${({ size, gutter }) => `
    height: ${getSizePixels(size, gutter)}px;
    width: ${getSizePixels(size, gutter)}px; 
  `}
`;

const CircularLoader = props => {
  const {
    size,
    skin,
    theme: {
      colors: {
        [skin]: { 700: color },
      },
      gutter,
    },
  } = props;

  return (
    <Wrapper size={size} gutter={gutter} role="progressbar">
      <Content color={color} viewBox="22 22 44 44">
        <Circle color={color} />
      </Content>
    </Wrapper>
  );
};

CircularLoader.defaultProps = {
  size: 'medium',
  skin: 'primary',
  theme: {
    colors,
    gutter: defaultGutter,
  },
};

CircularLoader.propTypes = {
  size: PropTypes.oneOf(['medium', 'large', 'xlarge', 'xxlarge', 'xxxlarge']),
  skin: PropTypes.oneOf(['primary', 'secondary']),
  /** Edit only for themification */
  theme: PropTypes.shape({
    colors: PropTypes.object,
    gutter: PropTypes.number,
  }),
};

export default CircularLoader;
