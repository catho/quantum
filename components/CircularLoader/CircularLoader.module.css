.wrapper {
    animation: circularRotate 1.4s linear infinite;
    position: relative;
}

.wrapper-medium {
    height: calc(var(--qtm-gutter) * 2);
    width: calc(var(--qtm-gutter) * 2);
}

.wrapper-large {
    height: calc(var(--qtm-gutter) * 3);
    width: calc(var(--qtm-gutter) * 3);
}

.wrapper-xlarge {
    height: calc(var(--qtm-gutter) * 4);
    width: calc(var(--qtm-gutter) * 4);
}

.wrapper-xxlarge {
    height: calc(var(--qtm-gutter) * 5);
    width: calc(var(--qtm-gutter) * 5);
}

.wrapper-xxxlarge {
    height: calc(var(--qtm-gutter) * 6);
    width: calc(var(--qtm-gutter) * 6);
}

.content {
    left: 50%;
    overflow: hidden;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    width: 100%;
    height: 100%;
}

.content-primary-color {
    color: var(--qtm-colors-primary-700);
}

.content-secondary-color {
    color: var(--qtm-colors-secondary-700);
}

.circle {
    animation: circularDash 1.4s ease-in-out infinite;
    cx: 44px;
    cy: 44px;
    fill: none;
    line-height: 1;
    r: 20.2px;
    stroke-dasharray: 80px, 200px;
    stroke-dashoffset: 0px;
    stroke-width: 3.6px;
    stroke: currentColor;
}

@keyframes circularRotate {
    100% {
        transform: rotate(360deg);
    }
}

@keyframes circularDash {
    0% {
        stroke-dasharray: 1px, 200px;
        stroke-dashoffset: 0px;
    }

    50% {
        stroke-dasharray: 100px, 200px;
        stroke-dashoffset: -15px;
    }

    100% {
        stroke-dasharray: 100px, 200px;
        stroke-dashoffset: -125px;
    }
}