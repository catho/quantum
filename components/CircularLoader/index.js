import CircularLoader from './CircularLoader';

export default CircularLoader;
