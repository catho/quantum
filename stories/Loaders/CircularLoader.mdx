import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import Header from '../shared/Header';
import CircularLoader from '../../components/CircularLoader';
import { Default, WithSkin, WithSize } from './CircularLoader.stories.jsx';

<Meta title="Loaders/CircularLoader" component={CircularLoader} />

<Header title="CircularLoader">
    CircularLoaders are used to give the user the impression of loading while a component is not ready to be displayed.
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Examples](#examples)

## <a id="importing"></a>Importing and usage

```js
import CircularLoader from '@catho/quantum/CircularLoader';
```

<Canvas>
  <Story story={Default}/>
</Canvas>

## <a id="api"></a>API

<ArgsTable of={CircularLoader} />

## <a id="examples"></a>Examples

### With skin

<Canvas>
  <Story story={WithSkin}/>
</Canvas>

### With size

<Canvas>
  <Story story={WithSize}/>
</Canvas>