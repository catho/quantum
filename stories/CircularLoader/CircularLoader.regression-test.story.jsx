import React from 'react';
import { storiesOf } from '@storybook/react';
import { CircularLoader } from '../../components';

storiesOf('CircularLoader', module)
  .add('default', () => <CircularLoader />)
  .add('with secondary theme', () => <CircularLoader skin="secondary" />)
  .add('with medium size', () => <CircularLoader size="medium" />)
  .add('with large size', () => <CircularLoader size="large" />)
  .add('with xlarge size', () => <CircularLoader size="xlarge" />)
  .add('with xxlarge size', () => <CircularLoader size="xxlarge" />)
  .add('with xxxlarge size', () => <CircularLoader size="xxxlarge" />);
