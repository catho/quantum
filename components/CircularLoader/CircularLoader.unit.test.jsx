import React from 'react';
import renderer from 'react-test-renderer';
import CircularLoader from './CircularLoader';

describe('CircularLoader component', () => {
  it('Should match the snapshot of a simple circular loader', () => {
    const simpleCircularLoader = <CircularLoader />;
    expect(renderer.create(simpleCircularLoader).toJSON()).toMatchSnapshot();
  });
});
