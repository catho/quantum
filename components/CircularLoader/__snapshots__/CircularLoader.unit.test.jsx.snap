// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularLoader component Should match the snapshot of a simple circular loader 1`] = `
.c1 {
  left: 50%;
  overflow: hidden;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  color: #1250C4;
}

.c2 {
  -webkit-animation: dPEoa 1.4s ease-in-out infinite;
  animation: dPEoa 1.4s ease-in-out infinite;
  cx: 44;
  cy: 44;
  fill: none;
  line-height: 1;
  r: 20.2;
  stroke-dasharray: 80px,200px;
  stroke-dashoffset: 0px;
  stroke-width: 3.6;
  stroke: currentColor;
  color: #1250C4;
}

.c0 {
  -webkit-animation: cZxgpV 1.4s linear infinite;
  animation: cZxgpV 1.4s linear infinite;
  position: relative;
  height: 16px;
  width: 16px;
}

<div
  className="c0"
  role="progressbar"
  size="medium"
>
  <svg
    className="c1"
    color="#1250C4"
    viewBox="22 22 44 44"
  >
    <circle
      className="c2"
      color="#1250C4"
    />
  </svg>
</div>
`;
